sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala

```diff
@@ -51,49 +51,46 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
       deviceBuffer,
       other.getSpillPriority,
       other.getSpillCallback)
   }
 
   /**
-   * Adds a contiguous table to the device storage, taking ownership of the table.
+   * Given a `DeviceMemoryBuffer` find out if a `MemoryBuffer.EventHandler` is associated
+   * with it. If so, make sure that it is an instance of `RapidsDeviceMemoryBuffer`.
    *
-   * @param table cudf table based from the contiguous buffer
-   * @param contigBuffer device memory buffer backing the table
-   * @param tableMeta metadata describing the buffer layout
-   * @param initialSpillPriority starting spill priority value for the buffer
-   * @param spillCallback a callback when the buffer is spilled. This should be very light weight.
-   *                      It should never allocate GPU memory and really just be used for metrics.
-   * @return RapidsBufferHandle handle for this table
+   * After getting the `RapidsDeviceMemoryBuffer` try and keep it alive via `addReference`.
+   * If successful, we can point to this buffer with a new handle, otherwise the buffer is
+   * about to be removed/freed (unlikely, because we are holding onto the reference as we
+   * are adding it again).
+   *
+   * @note public for testing
+   * @param buffer - the `DeviceMemoryBuffer` to inspect
+   * @return - Some(RapidsDeviceMemoryBuffer): the handler is associated with a rapids buffer
+   *         and the rapids buffer is currently valid, or
+   *
+   *         - None: if no `RapidsDeviceMemoryBuffer` is associated with this buffer (it is
+   *         brand new to the store, or the `RapidsDeviceMemoryBuffer` is invalid and
+   *         about to be removed).
    */
-  def addTable(
-      table: Table,
-      contigBuffer: DeviceMemoryBuffer,
-      tableMeta: TableMeta,
-      initialSpillPriority: Long,
-      spillCallback: SpillCallback = RapidsBuffer.defaultSpillCallback): RapidsBufferHandle = {
-    // We increment this because this rapids device memory has two pointers to the buffer:
-    // the actual contig buffer, and the table. When this `RapidsBuffer` releases its resources,
-    // it will decrement the ref count for the contig buffer (negating this incRefCount),
-    // it will also close the table being passed here, which together brings the ref count
-    // to 0.
-    contigBuffer.incRefCount()
-    val id = TempSpillBufferId()
-    freeOnExcept(
-      new RapidsDeviceMemoryBuffer(
-        id,
-        contigBuffer.getLength,
-        tableMeta,
-        Some(table),
-        contigBuffer,
-        initialSpillPriority,
-        spillCallback)) { buffer =>
-      logDebug(s"Adding table for: [id=$id, size=${buffer.size}, " +
-        s"meta_id=${buffer.meta.bufferMeta.id}, meta_size=${buffer.meta.bufferMeta.size}]")
-      addDeviceBuffer(buffer, needsSync = true)
-      catalog.makeNewHandle(id, initialSpillPriority, spillCallback)
+  def getExistingRapidsBufferAndAcquire(
+      buffer: DeviceMemoryBuffer): Option[RapidsDeviceMemoryBuffer] = {
```

Comment:
Do we need synchronization here to solve the case of multiple threads all referencing the same buffer and all deciding to make it spillable at the same time?

Reply:
Yeah this is similar to @revans2 comment. It was missed. 

It would be bad if 2 callers want to add the same buffer and they are added as separate RapidsDeviceMemoryBuffers. It would trigger a require I put into this class if the event handler was already set at runtime and that was not intended.